internal/common/rules/globalrulescache_test.go

```diff
@@ -13,6 +13,23 @@ func TestRegisterNewRule(t *testing.T) {
 	}
 }
 
+func TestPullRuleIntoPlay(t *testing.T) {
```

Comment:
I'd exercise all the conditions, and use a Test Table for convenience, for example:

```
cases := []struct{
    name string
    availRules []string
    rulesInPlay []string
    ruleToInsert string
    want error
}{
    {
        name: "normal working",
        ....,
    },
    {
        ....,
    },
}

for _, tc := range cases {
    t.Run(tc.name, func(t *testing.T) {
        got := PullRuleIntoPlay(tc.rule)
        testutils.CompareTestErrors(tc.want, got, t)
    })
}
```

See more examples in test cases I wrote.

Reply:
You'd also likely benefit from *Dependency Injection*, i.e. passing the dependencies as arguments to the function. This makes testing immensely easier. 

https://play.golang.org/p/ElKQf1-FlKs

Reply:
Oooo nice the test table thing looks good. I've put that in the code now. Thanks 👍